editor/src/components/inspector/fill-hug-fixed-control.tsx

```diff
@@ -286,371 +317,241 @@ export const FillHugFixedControl = React.memo<FillHugFixedControlProps>((props)
-          allElementPropsRef.current,
-          setPropFixedStrategies('always', 'horizontal', value),
-        )
-        return
-      }
-      if (widthCurrentValue.fixedHugFill?.type === 'fill') {
-        if (value.unit != null && value.unit !== '%') {
-          // fill mode only accepts percentage or valueless numbers
-          return
-        }
-        executeFirstApplicableStrategy(
-          dispatch,
-          metadataRef.current,
-          selectedViewsRef.current,
-          elementPathTreeRef.current,
-          allElementPropsRef.current,
-          setPropFillStrategies('horizontal', value.value, false),
+  const innerLabel = dimension.charAt(0).toUpperCase()
+
+  return (
+    <NumberInput
+      labelInner={innerLabel}
+      id={FillFixedHugControlId(dimension)}
+      testId={FillFixedHugControlId(dimension)}
+      value={fixedValue}
+      onSubmitValue={onAdjustValue}
+      onTransientSubmitValue={onAdjustValue}
+      onForcedSubmitValue={onAdjustValue}
+      controlStatus={inputControlStatus}
+      numberType={pickNumberType(currentValue.fixedHugFill)}
+      incrementControls={true}
+      stepSize={1}
+      minimum={0}
+      maximum={Infinity}
+      defaultUnitToHide={null}
+      focusOnMount={false}
+    />
+  )
+})
+WidthHeightNumberControl.displayName = 'WidthHeightNumberControl'
+
+function useOnSubmitFixedFillHugType(dimension: 'width' | 'height') {
+  const axis = dimension === 'width' ? 'horizontal' : 'vertical'
+
+  const dispatch = useDispatch()
+  const metadataRef = useRefEditorState(metadataSelector)
+  const selectedViewsRef = useRefEditorState(selectedViewsSelector)
+  const elementPathTreeRef = useRefEditorState((store) => store.editor.elementPathTree)
+  const allElementPropsRef = useRefEditorState((store) => store.editor.allElementProps)
+
+  const currentValueRef = useComputedSizeRef(dimension)
+
+  const fillsOtherAxisRef = useRefEditorState(
+    (store) =>
+      detectFillHugFixedStateMultiselect(
+        axis === 'vertical' ? 'horizontal' : 'vertical',
+        metadataSelector(store),
+        selectedViewsSelector(store),
+      ).fixedHugFill?.type === 'fill',
+  )
+
+  const onSubmitFixedFillHugType = React.useMemo(() => {
+    return ({ value: anyValue }: SelectOption) => {
+      const value = anyValue as FixedHugFillMode
+
+      const currentComputedValue = currentValueRef.current ?? 0
+
+      const firstSelectedView = safeIndex(selectedViewsRef.current, 0)
+      if (firstSelectedView != null) {
+        const valueToUse =
+          dimension === 'width'
+            ? fixedSizeDimensionHandlingText(
+                metadataRef.current,
+                elementPathTreeRef.current,
+                firstSelectedView,
+                currentComputedValue,
+              )
+            : currentComputedValue
+        const strategy = strategyForChangingFillFixedHugType(
+          valueToUse,
+          dimension === 'width' ? 'horizontal' : 'vertical',
```

Comment:
Fairly certain you can use this.
```suggestion
          axis,
```